test/extensions/filters/http/lua/wrappers_test.cc

```diff
@@ -250,6 +260,138 @@ TEST_F(LuaRequestInfoWrapperTest, ReturnCurrentProtocol) {
   expectToPrintCurrentProtocol(Http::Protocol::Http2);
 }
 
+// Set, get and iterate request info dynamic metadata.
+TEST_F(LuaRequestInfoWrapperTest, SetGetAndIterateDynamicMetadata) {
+  const std::string SCRIPT{R"EOF(
+      function callMe(object)
+        testPrint(type(object:dynamicMetadata()))
+        object:dynamicMetadata():set("envoy.lb", "foo", "bar")
+        object:dynamicMetadata():set("envoy.lb", "so", "cool")
+
+        testPrint(object:dynamicMetadata():get("envoy.lb")["foo"])
+        testPrint(object:dynamicMetadata():get("envoy.lb")["so"])
+
+        for filter, entry in pairs(object:dynamicMetadata()) do
+          for key, value in pairs(entry) do
+            testPrint(string.format("'%s' '%s'", key, value))
+          end
+        end
+
+        local function nRetVals(...)
+          return select('#',...)
+        end
+        testPrint(tostring(nRetVals(object:dynamicMetadata():get("envoy.ngx"))))
+      end
+    )EOF"};
+
+  InSequence s;
+  setup(SCRIPT);
+
+  RequestInfo::RequestInfoImpl request_info(Http::Protocol::Http2);
+  EXPECT_EQ(0, request_info.dynamicMetadata().filter_metadata_size());
+  Filters::Common::Lua::LuaDeathRef<RequestInfoWrapper> wrapper(
+      RequestInfoWrapper::create(coroutine_->luaState(), request_info), true);
+  EXPECT_CALL(*this, testPrint("userdata"));
+  EXPECT_CALL(*this, testPrint("bar"));
+  EXPECT_CALL(*this, testPrint("cool"));
+  EXPECT_CALL(*this, testPrint("'foo' 'bar'"));
+  EXPECT_CALL(*this, testPrint("'so' 'cool'"));
+  EXPECT_CALL(*this, testPrint("0"));
+  start("callMe");
+
+  EXPECT_EQ(1, request_info.dynamicMetadata().filter_metadata_size());
+  EXPECT_EQ("bar", request_info.dynamicMetadata()
+                       .filter_metadata()
+                       .at("envoy.lb")
+                       .fields()
+                       .at("foo")
+                       .string_value());
+  wrapper.reset();
+}
+
+// Modify during iteration.
```

Comment:
I'm still not convinced this is safe. Looking quickly here: https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.map it's not clear from these docs when the iterator is invalidated, but if it's anything like `unordered_map` I think that iterators can get invalidated sometimes: https://stackoverflow.com/questions/16781886/can-we-store-unordered-maptiterator. Can you maybe poke around the proto map implementation to understand whether this safe in all cases and guaranteed to be safe in the future? I'm mainly wondering if it would be safer/easier to block modification during iteration like we do for the header map currently?

Reply:
> I'm mainly wondering if it would be safer/easier to block modification during iteration like we do for the header map currently?

One question/clarification, do you mean by throwing an error like the following?

https://github.com/dio/envoy/blob/36f01b8c35c9b3decff7c43ef2ba8e5e2ee0e2ed/source/extensions/filters/http/lua/wrappers.cc#L153-L155

Reply:
Yeah that's what I was referring to. If it's safe in all cases is fine, I'm just not sure it is. Up to you whether you want to block or investigate more.

Reply:
Got it, I will explore more then.

Reply:
@mattklein123 based on this: https://github.com/google/protobuf/blob/88795f64e4c5dc870c63b3d4d39924cc91590b2f/src/google/protobuf/map.h#L326-L327

```
// 8. Mutations to a map do not invalidate the map's iterators, pointers to
//    elements, or references to elements.
```

It seems the iterator is saved during mutation/rehash. Does it mean we can allow mutation during iteration?

Reply:
Yeah seems OK. What if someone deletes the element that `current_` points to though? Can you test that? In your implementation I guess it would be the element after the one you are about to return via the __pairs function?

Reply:
I'm not sure I'm following. Do you mean deletion from Lua? 

Since we don't have that API yet (only `set` and `get` for now, I think I will try to set `current_` to point to something else, i.e. a struct with the same key but a different value will give the same effect?).

Reply:
FWIW, I'm thinking to just block any mutation in mutation API like `set` as your suggestion just to be safe.

Reply:
What I meant is right now when you iterate in the __pairs function, you set `current_` to the next set member before you return the value to Lua. What if Lua then deletes that next member? What happens to the iterator? I have to think that it's no longer valid and bad things will happen? 

Yeah, TBH, I think for now we should just do the block so we don't have to keep thinking about all of these edge cases. Is that already implemented or do you want to change it?

Reply:
> What I meant is right now when you iterate ...

Thank you for this. This really makes me think about this all. Should be careful next time.

> I think for now we should just do the block

If I understand you correctly, 

I have the following before the setting a metadata entry. https://github.com/envoyproxy/envoy/pull/3800/files#diff-cb2388822be823ac2236f97bc5cc776dR153

```
if (iterator_.get() != nullptr) {
  luaL_error(state, "dynamic metadata map cannot be modified while iterating");
}
```

Do I need to check somewhere else? Thanks.

Reply:
No I think that's fine. Will review again.